src/DiskIO/IpcIo/IpcIoFile.cc

```diff
@@ -868,6 +877,17 @@ IpcIoFile::WaitBeforePop()
     return false;
 }
 
+pid_t
+IpcIoFile::getPid()
+{
+    static auto pid = getpid();
+    // pid may become wrong if we fork(). The assertion below guards us from the problem
+    // because currently no worker or disker forks into another worker or disker.
+    // However, this may change in future. TODO: add a more reliable check instead.
+    assert(TheProcessKind == pkWorker || TheProcessKind == pkDisker);
```

Comment:
Initially, I tried to implement the approach we discussed earlier, i.e., add a global 'fork' counter and check it here. However, I faced some difficulties and rejected it.

forkCounter and fork() can become unsynchronized if someone adds fork() and forgets to care about forkCounter(). Adding a fork() wrapper would fully protect from this because they are still free to call fork() somewhere. If we add the wrapper anyway, this would not be a trivial fork() -> ourForkWrapper() change, because probably we need to return some additional parameters, e.g., errno (which should be saved immediately after fork()). Also, there are additional difficulties with Windows support (e.g., ipcCreate() in ipc_win32.cc) - where we also need to increment forkCounter.

Fortunately, there are few fork() calls in Squid, and some of them are followed by exec() methods, overriding the image of the child process. So there are only two fork() without exec() which we should care about - in ipcCreate() from ipc.cc and GoIntoBackground() from main.cc. In the first case, the child process kind is pkOther and in the second - pkHelper. So I think that with the above TheProcessKind assertion we can do without forkCounter (at least for now).